pkg/pillar/cmd/nodeagent/handletimers.go

```diff
@@ -0,0 +1,188 @@
+	if !ctx.updateInprogress || ctx.testInprogress {
+		return
+	}
+	// only when current partition state is in progress
+	// start the test
+	ctx.upgradeTestStartTime = time.Now()
+	ctx.testInprogress = true
+	successLimit := time.Second *
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	ctx.remainingTestTime = successLimit
+	ctx.testInprogress = true
+}
+
+func resetTestStartTime(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	// the led status has changed, incase, the test is in progress,
+	// we need to start the test again, reset first
+	ctx.testInprogress = false
+	ctx.remainingTestTime = 0
+	ctx.testInprogress = false
+}
+
+func setConfigGetFailState(ctx *nodeagentContext) {
+	if !ctx.configGetFail {
+		ctx.configGetFail = true
+	}
+}
+
+func resetConfigGetFailState(ctx *nodeagentContext) {
+	if ctx.configGetFail {
+		ctx.configGetFail = false
+		ctx.configGetFailCount = 0
+	}
+}
+
+func executeConfigGetFailState(ctx *nodeagentContext) {
+	if !ctx.configGetFail {
+		return
+	}
+	if ctx.configGetFailCount >= maxConfigGetFailCount {
+		resetConfigGetFailState(ctx)
+		resetTestStartTime(ctx)
+	}
+}
+
+func checkUpgradeValidationTestTimeExpiry(ctx *nodeagentContext) bool {
+	if !ctx.testInprogress {
+		return false
+	}
+	timePassed := time.Since(ctx.upgradeTestStartTime)
+	successLimit := time.Second *
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	if timePassed < successLimit {
+		ctx.remainingTestTime = successLimit - timePassed
+		log.Infof("CurPart: %s inprogress, waiting for %d seconds\n",
+			ctx.curPart, ctx.remainingTestTime/time.Second)
+		publishNodeAgentStatus(ctx)
+		return false
+	}
+	return true
+}
+
+func handleConfigGetFail(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	if !ctx.configGetFail {
+		ctx.configGetFail = true
+	}
+	ctx.configGetFailCount++
+	executeConfigGetFailState(ctx)
+}
+
+func updateLastConfigReceivedTime(ctx *nodeagentContext,
+	status types.ZedAgentStatus) {
+	log.Debugf("lastConfigReceivedtime: %v, failCount: %d \n",
+		status.LastConfigReceivedTime, status.ConfigGetFailCount)
+	// time stamp has not changed, since last config receive
+	if ctx.lastConfigReceivedTime == status.LastConfigReceivedTime {
+		handleConfigGetFail(ctx)
+		return
+	}
+	// config get is successful
+	resetConfigGetFailState(ctx)
+	setTestStartTime(ctx)
+	ctx.lastConfigReceivedTime = status.LastConfigReceivedTime
+	if checkUpgradeValidationTestTimeExpiry(ctx) {
+		log.Infof("CurPart: %s, Upgrade Validation Test Complete\n",
+			ctx.curPart)
+		resetTestStartTime(ctx)
+		initiateBaseOsZedCloudTestComplete(ctx)
+		publishNodeAgentStatus(ctx)
+	}
+}
+
+func execReboot(ctx *nodeagentContext, rebootStr string) {
+
+	log.Infof("execReboot(): Reboot reason(%s)\n", rebootStr)
+	if ctx.needsReboot {
+		log.Infof("reboot flag is already set\n")
+		return
+	}
+
+	ctx.needsReboot = true
+	ctx.rebootReason = rebootStr
+	publishNodeAgentStatus(ctx)
+
+	// if zedagent is not alive, handle it here
```

Comment:
Why not consistently handle the failure reboot in one place? Makes for less test cases and less bug opportunities.

Reply:
When the zedagent is active, the idea is to gracefully shutdown the application instances, as zedagent is the publisher of application instance config. Otherwise, it will get executed in the nodeagent context itself.